KAFKA-15922: Bump MetadataVersion to support JBOD with KRaft

pprovenzano opened the pull request:
Update the LATEST_PRODUCTION to include JBOD and fetch request for KIP-951. 

Remove ELR MetadataVersion from 3.7 and push it to 3.8. Since it was not considered stable the MV is burned.

Add 3.8-IV0 as the next MetadataVersion which is currently not considered stable.

Fix tests to reflect this state.

pprovenzano committed:
Bump MetadataVersion to support JBOD with KRaft

pprovenzano committed:
All Metadata versions post JBOD are not in 4.0.
This just moved 3.7_IV3 to 4.0_IV0

rondagostino reviewed (approved):
LGTM assuming tests pass.

OmniaGM reviewed (approved):
LGTM

rondagostino commented:
@pprovenzano Can you look at the test failures?
`60 tests have failed. There are 0 new tests failing, 60 existing failing`
https://ci-builds.apache.org/blue/organizations/jenkins/Kafka%2Fkafka-pr/detail/PR-14984/2/tests

pprovenzano committed:
We aren't going to support migration in 4.0 because ZK won't be in 4.0
There is no short version of 4.0 until 4.0 becomes a stable version

pprovenzano commented:
Okay, I think I have addresses the issues.

ijuma commented on server-common/src/test/java/org/apache/kafka/server/common/MetadataVersionTest.java:
```diff
@@ -166,13 +166,12 @@ public void testFromVersionString() {
         assertEquals(IBP_3_6_IV2, MetadataVersion.fromVersionString("3.6-IV2"));
 
         // 3.7-IV0 is the latest production version in the 3.7 line
```
This comment doesn't seem correct.

pprovenzano replied:
Thank you!
Fixed

pprovenzano committed:
Update comment

pprovenzano committed:
Another missed test fixed.

pprovenzano committed:
Update MVs to the agreed upon resolution:

3.7-IV0 (15): Controller Registration
3.7-IV1 (16): burned in AK, reused in CE for CL fail-back
3.7-IV2 (17): JBOD
3.7-IV4 (19): Fetch RPC change (<-- “Production Ready”)
3.8-IV0 (20): ELR

MVs of non Production Ready items cannot be recyled but the feature itself can be removed and
consume a later MV. This is the case with ELR which is no longer 3-7-IV3 (18) and has been
moved to 3-8-IV0 which is not production ready.

rondagostino commented on core/src/test/scala/integration/kafka/zk/ZkMigrationIntegrationTest.scala:
```diff
@@ -74,7 +74,8 @@ object ZkMigrationIntegrationTest {
       MetadataVersion.IBP_3_7_IV1,
       MetadataVersion.IBP_3_7_IV2,
-      MetadataVersion.IBP_3_7_IV3
+      MetadataVersion.IBP_3_7_IV4,
+      MetadataVersion.IBP_3_8_IV0
```
Notng that we are removing MetadataVersion.IBP_3_7_IV3 because it is "burned" and will never be used.

rondagostino replied:
Maybe we should keep this here as per the other comments?  Not sure, since the test will take some time, and maybe we don't want to spend it.

pprovenzano replied:
I'm against adding it here as it doesn't add anything. IV4 and IV2 are the important ones.

rondagostino commented on server-common/src/main/java/org/apache/kafka/server/common/MetadataVersion.java:
```diff
@@ -194,8 +194,13 @@ public enum MetadataVersion {
     // Add JBOD support for KRaft.
     IBP_3_7_IV2(17, "3.7", "IV2", true),
 
+    // IBP_3_7_IV3 was ELR related supports (KIP-966) and has been moved forward
```
I think we should keep the value and just not have anything associated with it.  Call it "reserved", similar to IBP_3_7_IV1.  WDYT?

jolshan replied:
I think that makes sense.

pprovenzano replied:
Done!

jolshan replied:
very small nit here -- but the comment could be a little clearer. Something like

IBP_3_7_IV3 was reserved for ELR changes (KIP-966) but has been deferred to a later MV

pprovenzano replied:
Okay, I updated the comment to try to make it clearer. This is where we really need the KIP.

rondagostino commented on server-common/src/test/java/org/apache/kafka/server/common/MetadataVersionTest.java:
```diff
@@ -165,13 +165,15 @@ public void testFromVersionString() {
 
-        // 3.7-IV0 is the latest production version in the 3.7 line
-        assertEquals(IBP_3_7_IV0, MetadataVersion.fromVersionString("3.7"));
+        // 3.7-IV2 is the latest production version in the 3.7 line
+        assertEquals(IBP_3_7_IV2, MetadataVersion.fromVersionString("3.7"));
```
Isn't it IBP_3_7_IV4?

pprovenzano replied:
Fixed!

rondagostino commented on server-common/src/test/java/org/apache/kafka/server/common/MetadataVersionTest.java:
```diff
@@ -165,13 +165,15 @@ public void testFromVersionString() {
         assertEquals(IBP_3_7_IV0, MetadataVersion.fromVersionString("3.7-IV0"));
         assertEquals(IBP_3_7_IV1, MetadataVersion.fromVersionString("3.7-IV1"));
         assertEquals(IBP_3_7_IV2, MetadataVersion.fromVersionString("3.7-IV2"));
-        assertEquals(IBP_3_7_IV3, MetadataVersion.fromVersionString("3.7-IV3"));
```
Keep it?

pprovenzano replied:
Done!

rondagostino commented on server-common/src/test/java/org/apache/kafka/server/common/MetadataVersionTest.java:
```diff
@@ -228,6 +230,8 @@ public void testShortVersion() {
         assertEquals("3.6", IBP_3_6_IV2.shortVersion());
         assertEquals("3.7", IBP_3_7_IV0.shortVersion());
         assertEquals("3.7", IBP_3_7_IV1.shortVersion());
+        assertEquals("3.7", IBP_3_7_IV2.shortVersion());
+        assertEquals("3.7", IBP_3_7_IV4.shortVersion());
```
Keep IV3?

pprovenzano replied:
Done

rondagostino commented on server-common/src/test/java/org/apache/kafka/server/common/MetadataVersionTest.java:
```diff
@@ -274,7 +278,8 @@ public void testVersion() {
         assertEquals("3.7-IV0", IBP_3_7_IV0.version());
         assertEquals("3.7-IV1", IBP_3_7_IV1.version());
         assertEquals("3.7-IV2", IBP_3_7_IV2.version());
-        assertEquals("3.7-IV3", IBP_3_7_IV3.version());
```
keep it?

pprovenzano replied:
Done!

rondagostino reviewed (commented):
Left a few comments based on latest changes.

pprovenzano committed:
Fixup with comments from Ron

pprovenzano committed:
Missed a test

jolshan reviewed (approved):
left one small nit, but otherwise lgtm assuming tests look good.

jsancio commented on server-common/src/main/java/org/apache/kafka/server/common/MetadataVersion.java:
```diff
@@ -194,8 +194,14 @@ public enum MetadataVersion {
     IBP_3_7_IV2(17, "3.7", "IV2", true),
 
+    // IBP_3_7_IV3 was ELR related supports (KIP-966) and has been moved forward
+    IBP_3_7_IV3(18, "3.7", "IV3", false),
+
+    // Add new fetch request version for KIP-951
+    IBP_3_7_IV4(19, "3.7", "IV4", false),
+
     // Add ELR related supports (KIP-966).
-    IBP_3_7_IV3(18, "3.7", "IV3", true);
+    IBP_3_8_IV0(20, "3.8", "IV0", true);
```
Can you split this change into two PRs/commits? One that fixes the MV in 3.7 and one that introduced the new 3.8 MV. The one that fixes the 3.7 versions should be cherry picked to the 3.7 branch.

pprovenzano replied:
If we remove the tag from here should we also remove all the other code associated with ELR from the 3.7 branch? 
I don't think removing it from the 3.7 branch really buys us much of anything except more code churn.

jsancio replied:
You just need to return `false` for `isElrSupported()`. Is that what you are asking?

pprovenzano replied:
No, you asked to split this PR into two separate PRs, one for 3.7 without any reference to 3.8 and this one. I am saying that doing that has no value because 3.8 is labeled as unstable so having it in 3.7 isn't an issue. Customers cannot use 3.8 unless they explicitly allow unstable metadata versions. 
If we want 3.7 to be clean then you would also need to remove all ELR code not just the MV referencing it. This would be a BIG change because it removes tests and updates the JSON files representing the record format.

jsancio reviewed (commented):
Please update the PR description to explain what issue is being fixed. The description doesn't match the changes included in this PR.

Same with the PR title. Is the KAFKA-15922 also related to this PR?

pprovenzano committed:
Update comment to try to make it a little clearer

pprovenzano commented:
> Please update the PR description to explain what issue is being fixed. The description doesn't match the changes included in this PR.
> 
> Same with the PR title. Is the [KAFKA-15922](https://issues.apache.org/jira/browse/KAFKA-15922) also related to this PR?

The original PRs for KAFKA-15922 are what got us to this point. My PR here was just meant to bump the latest stable version to the MV supporting JBOD. I've added a little more to reduce the number of PRs that need to be committed to streamline things. All the changes though are related to KAFKA-15922. I have updated the comment to try to reflect the current state of changes. I have no idea what to change the title to.

ijuma commented:
Java 21 build passed, others have failures that look unrelated:

> Build / JDK 11 and Scala 2.13 / org.apache.kafka.connect.mirror.integration.MirrorConnectorsIntegrationBaseTest.testSyncTopicConfigs()
Build / JDK 11 and Scala 2.13 / kafka.network.ConnectionQuotasTest.testListenerConnectionRateLimitWhenActualRateAboveLimit()
Build / JDK 11 and Scala 2.13 / kafka.network.SocketServerTest.testSaslReauthenticationFailureWithKip152SaslAuthenticate()
Build / JDK 11 and Scala 2.13 / org.apache.kafka.trogdor.coordinator.CoordinatorTest.testTaskRequestWithOldStartMsGetsUpdated()
Build / JDK 17 and Scala 2.13 / kafka.api.ConsumerBounceTest.testConsumptionWithBrokerFailures()
Build / JDK 17 and Scala 2.13 / org.apache.kafka.tiered.storage.integration.TransactionsWithTieredStoreTest.testBumpTransactionalEpoch(String).quorum=kraft
Build / JDK 8 and Scala 2.12 / org.apache.kafka.connect.mirror.integration.MirrorConnectorsIntegrationTransactionsTest.testSyncTopicConfigs()
Build / JDK 8 and Scala 2.12 / kafka.api.DelegationTokenEndToEndAuthorizationWithOwnerTest.testProduceConsumeWithWildcardAcls(String).quorum=kraft
Build / JDK 8 and Scala 2.12 / kafka.api.SaslMultiMechanismConsumerTest.testCoordinatorFailover(String, String).quorum=zk.groupProtocol=generic
Build / JDK 8 and Scala 2.12 / kafka.network.ConnectionQuotasTest.testListenerConnectionRateLimitWhenActualRateAboveLimit()
Build / JDK 8 and Scala 2.12 / org.apache.kafka.controller.QuorumControllerTest.testBalancePartitionLeaders()
Build / JDK 8 and Scala 2.12 / org.apache.kafka.tiered.storage.integration.TransactionsWithTieredStoreTest.testBumpTransactionalEpoch(String).quorum=kraft
Build / JDK 8 and Scala 2.12 / org.apache.kafka.tiered.storage.integration.TransactionsWithTieredStoreTest.testSendOffsetsToTransactionTimeout(String).quorum=kraft
Build / JDK 8 and Scala 2.12 / org.apache.kafka.tools.MetadataQuorumCommandTest.testDescribeQuorumStatusSuccessful [5] Type=Raft-Combined, MetadataVersion=3.8-IV0, Security=PLAINTEXT
Build / JDK 11 and Scala 2.13 / kafka.api.TransactionsBounceTest.testWithGroupId()

rondagostino commented on server-common/src/main/java/org/apache/kafka/server/common/MetadataVersion.java:
```diff
@@ -393,7 +400,7 @@ public short partitionRecordVersion() {
     }
 
     public short fetchRequestVersion() {
-        if (this.isAtLeast(IBP_3_7_IV0)) {
+        if (this.isAtLeast(IBP_3_7_IV4)) {
```
Noting that we do not have a test for this.  In the interest of time, let's add a test for this in a follow-on PR rather than here unless we find other issues.

rondagostino commented on tools/src/test/java/org/apache/kafka/tools/FeatureCommandTest.java:
```diff
@@ -68,17 +68,17 @@ public void testDescribeWithKRaft() {
     }
 
-    @ClusterTest(clusterType = Type.KRAFT, metadataVersion = MetadataVersion.IBP_3_7_IV0)
+    @ClusterTest(clusterType = Type.KRAFT, metadataVersion = MetadataVersion.IBP_3_7_IV4)
```
Seem that this change is unnecessary, but it doesn't hurt, either - so let's leave it alone in the interest of time unless we find something else that definitely needs fixing.

rondagostino reviewed (approved):
LGTM.  One missing test that we can add separately (it was always missing -- it is not missing due to this PR). I will merge this and then cherry-pick it to 3.7.  Thanks for the adjustments!

rondagostino commented:
Merged to 3.7
